### PR TITLE
Improve constant name lookup

### PR DESCRIPTION
**Pull request by paracycle**
There were a few things wrong with the way the current code was looking up constant names:

1. The parameter supplied to the `path` method can be typed as a `Module` since `Class` is a subclass of `Module`, thus, every constant would be a `Module`.
2. The parameter name as `object` was misleading, since we don't expect any random object, but a constant.
3. Looking up the name of a constant via the `to_s` method is a world of pain, since `to_s` does more magic like converting singleton names to `#<Class:Foo>` format or refinement names to `#<refinement:Bar xxx>` format. All of these are ultimately useless as a class/module identifier and cannot be processed properly by this method. We should be looking up the actual `name` of the constant.
4. Looking up the actual name of the constant is not straight-forward since a lot of code in the wild actually override the `name` class method and return non-sensical values. What gives the most accurate result is to call the `Module.name` method on the current constant instance, which can be achieved using the `bind.call` trick. This ensures that we get the identifier used to originally bind the constant with, not a value returned from a potentially overridden `name` method.

This PR adds tests and makes code changes to ensure that `Namespace#path` is safe against these pitfalls

### CODE REVIEW

File: lib/parlour/rbi_generator/namespace.rb [hunk 1]
```diff
@@ -106,18 +106,24 @@ def constants
         )
       end
 
-      sig { params(object: T.untyped, block: T.proc.params(x: Namespace).void).void }
-      # Given a Class or Module object, generates all classes and modules in the
+      sig { params(constant: Module, block: T.proc.params(x: Namespace).void).void }
+      # Given a Module object, generates all classes and modules in the
```
Comment (Morriar) [inline lib/parlour/rbi_generator/namespace.rb:110]: Good catch, let's also update the comment in that case

Comment (Morriar) [reply lib/parlour/rbi_generator/namespace.rb:110]: I'm not sure how but `constant` should appear somewhere in the sentence to make the link with `Module object`

Comment (paracycle) [reply lib/parlour/rbi_generator/namespace.rb:110]: I can reword it to: "Given a constant (i.e. a Module instance), generates ..."

How does that sound?

Comment (Morriar) [reply lib/parlour/rbi_generator/namespace.rb:110]: Yeah I understand the comment better, thanks! 👍

**Comment by AaronC81** [comment]
Looks great, thanks for your contribution @paracycle!

I agree with the rewording of the docs in the review, so if you wouldn't mind committing that please, then I'll get this merged in.

**Comment by paracycle** [comment]
> Looks great, thanks for your contribution @paracycle!
> 
> I agree with the rewording of the docs in the review, so if you wouldn't mind committing that please, then I'll get this merged in.

Thank you. Sorry for the delay but I was only able to get the wording change in now.

BTW, there is a similar problem with calling `.class` when looking up the classes of the name parts, since it is not uncommon for code to override `Kernel.class` either. I can make a separate PR to fix that as well, if you'd like.

**Comment by AaronC81** [comment]
Cheers @paracycle - if you think that using `.class` could be an issue, then a PR would be very much appreciated :D
